crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -789,6 +799,18 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    fn infer_match_pattern_definition(
+        &mut self,
+        _pattern: &ast::Pattern,
+        _index: u32,
+        definition: Definition<'db>,
+    ) {
+        // TODO(dhruvmanila): The correct way to infer types here is to perform structural matching
+        // against the subject expression type (available in `self.index.expression_ty(subject)`)
+        // and extract the type at the `index` position if the pattern matches.
+        self.types.definitions.insert(definition, Type::Unknown);
+    }
+
     fn infer_match_pattern(&mut self, pattern: &ast::Pattern) {
```

Comment:
In this method, anywhere we would have created a `Definition` in semantic indexing, we should be calling `self.infer_definition(...)` to lookup the Definition by node (identifier in this case) and infer types in that definition's region. This is what ensures we never double infer types: a given expression or definition is only inferred in one region, and "outer" regions always get access to those types by going through the cached Salsa query.

At least that's the design so far. This case is a little odd, because a match pattern definition is a very limited region (just an identifier) that doesn't actually include any expressions. Currently there is also an invariant that we maintain that doing scope-level inference will infer a type for every Definition in that scope, and have those types in its `self.types.definitions` map. But I actually don't think we rely on that invariant anywhere, like we do on the all-expressions-have-types invariant.

I still think for now it's probably best if we go ahead and do inference on all the match-pattern Definitions when we encounter them here? 

Looking at this now, I also think when we do proper type inference for match statements, we will need to create a new ingredient and Salsa query for inferring types on a pattern and matching it against a subject, otherwise we will end up inferring types on expressions within the pattern multiple times (once per definition in that pattern), which is something we want to avoid. That new query doesn't need to happen in this PR, but it would be good to add a TODO here, something like `// TODO Salsa query for inferring pattern types and matching against subject`.

Reply:
> I still think for now it's probably best if we go ahead and do inference on all the match-pattern Definitions when we encounter them here?

This requires creating the definition which requires the identifier index in the root pattern. This means we need to use the visitor pattern / save context on the type inference builder. I avoided this for now because I wasn't sure exactly how the type inference was going to work here which itself is a separate task. Still, if you want I can take that up as a follow-up work to infer the definition types by storing the context on the builder.

Reply:
I'm not sure I understand "This requires creating the definition which requires the identifier index in the root pattern" -- I wasn't suggesting to fully do the inference here, just to actually make it go through the `infer_definition_types` query path (but then still end up with `Unknown` for now). But now that this is merged it's fine, we can follow up in a future PR on the inference part.

Reply:
I can look at it tomorrow morning. I might need to go through the discussion again to expand my comment ;)